[DOCS-2246] Create session-replay-getting-started.md (#11066)
DataDog/documentation: content/en/real_user_monitoring/guide/session-replay-getting-started.md

---
title: Getting started with Session Replay
kind: guide
description: Guide for enabling Session Replay and how to set privacy options
beta: true
further_reading:
- link: '/real_user_monitoring/explorer'
  tag: 'Documentation'
  text: 'Visualize your RUM data in the Explorer'
---

## What is Session Replay?

Session Replay expands your UX monitoring by allowing you to capture and visually replay the web browsing experience of your users.
Used in combination with RUM performance data, Session Replay can be an asset for error identification, reproduction & resolution, and can bring invaluable insights into your web application’s usage patterns and design pitfalls.

## Collect Session Replay data

### Initial setup

To use Session Replay, Datadog RUM Browser Monitoring must already be set up, following the steps detailed [here][1]: application creation, client token generation, and RUM SDK setup.

> Session Replay is available through a dedicated build of the RUM Browser SDK. As such, enabling Session Replay only requires changing the npm package name or CDN URL, depending on your chosen installation method.

Author: npm is always lowercase

#### npm

> The `@datadog/browser-rum package` is to be replaced with `[@datadog/browser-rum-recorder][2]` When `datadogRum.init()` is called, the Session Replay recording is started as well.

Reviewer: ```suggestion
Replace the `@datadog/browser-rum package` with `[@datadog/browser-rum-recorder][2]` When `datadogRum.init()` is called, it also starts the Session Replay recording.
```

``` javascript
import { datadogRum } from '@datadog/browser-rum-recorder'

datadogRum.init({
  applicationId: '<DATADOG_APPLICATION_ID>',
  clientToken: '<DATADOG_CLIENT_TOKEN>',
  site: '<DATADOG_SITE>',
  //  service: 'my-web-application',
  //  env: 'production',
  //  version: '1.0.0',
  sampleRate: 100,
  trackInteractions: true,
})
```

#### CDN

> The Browser SDK URL https://www.datadoghq-browser-agent.com/datadog-rum.js is to be replaced with https://www.datadoghq-browser-agent.com/datadog-rum-recorder.js When `DD_RUM.init()` is called, the Session Replay recording is started as well.

Reviewer: ```suggestion
Replace the Browser SDK URL `https://www.datadoghq-browser-agent.com/datadog-rum.js` with `https://www.datadoghq-browser-agent.com/datadog-rum-recorder.js`. When `DD_RUM.init()` is called, it also starts the Session Replay recording.
```

> *Supported browsers*: The Session Replay recorder supports all the browsers supported by the RUM Browser SDK with the exception of IE11. See the [browser support table][3].

Reviewer: ```suggestion
*Supported browsers*: The Session Replay recorder supports all the browsers supported by the RUM Browser SDK with the exception of IE11. For a full list of supported browsers, see the [browser support table][3].
```
it might also be good to update this linked table to note that IE11 does not support Session Replay.

#### Configuration

The usual [RUM initialization parameters][4] are all supported. 

> In addition, the Session Replay recorder accepts a `manualSessionReplayRecordingStart` option. If set to `true`, the Session Replay recording will not start automatically when calling `init()`. To then start the recording, call `startSessionReplayRecording()`. This can be useful to conditionally start the recording (e.g, to only record authenticated user sessions):

Reviewer: ```suggestion
In addition, the Session Replay recorder accepts a `manualSessionReplayRecordingStart` option. If set to `true`, the Session Replay recording does not start automatically when calling `init()`. To then start the recording, call `startSessionReplayRecording()`. This can be useful to conditionally start the recording (for example to only record authenticated user sessions):
```

``` javascript
DD_RUM.init({
	manualSessionReplayRecordingStart: true
})
if (user.isAuthenticated) {
	DD_RUM.startSessionReplayRecording()
}
```

> The Session Replay recording can be stopped by calling `stopSessionReplayRecording()`.

Reviewer: ```suggestion
Stop the Session Replay recording by calling `stopSessionReplayRecording()`.
```

### Sensitive & personal data obfuscation

> By default, any input events on HTML input elements of type `password`, `email` and `tel` are ignored.

Reviewer: ```suggestion
By default, any input events on HTML input elements of type `password`, `email`, and `tel` are ignored.
```

``` html
<form method="post" data-dd-privacy=”input-ignored”>
    <input type="text" name="name" id="name" required>
    <input type="number" name="age" id="age" required>
    <input type="email" name="email" id="email" required>
    <input type="submit" value="submit">
</form>
```

Additionally, HTML elements can be fully obfuscated. These elements are replaced with a white block at recording time, as seen in this example of the navbar of a replay of a Datadog session: 

{{< img src="real_user_monitoring/guide/replay-hidden.png" alt="Replay Hidden Example">}}

Obfuscate elements by setting the data attribute `data-dd-privacy` to hidden, or by adding a class of `dd-privacy-hidden`. Elements marked as hidden will *not* have their content recorded and the PII will not be sent to Datadog. For example, the following div will be obfuscated and replaced in the replay by a white block of the same size:

``` html
<div id=”profile-info” data-dd-privacy=”hidden”>
    <p>Name: John Doe</p>
    <p>Birth date: June 6th, 1987</p>
</div>
```

## Troubleshooting

### Some HTML elements are not visible at replay

> Session Replay does not support the following HTML elements at the moment: `iframe`, `video`, `audio`, `canvas`, as well as Web Components.

Reviewer: ```suggestion
Session Replay does not support the following HTML elements: `iframe`, `video`, `audio`, `canvas`, and web components.
```

### Fonts or images do not render properly

> A Session Replay is not video, but an actual iframe rebuilt based on snapshots of the DOM. The replay thus depends on the various assets of the page: fonts & images.

Reviewer: ```suggestion
A Session Replay is not video, but an actual iframe rebuilt based on snapshots of the DOM. The replay depends on the various assets of the page: fonts & images.
```

Several reasons might explain why assets are not available at the time of the replay.

> 1. The resource does not exist anymore. For example, it was part of a previous deployment.

Reviewer: ```suggestion
- The resource does not exist anymore. For example, it was part of a previous deployment.
```
changing this to a bulleted list because we only use an ordered list if you need to complete steps in the specific order

> 2. The resource is inaccessible. For example, authentication might be required, or the resource might only be accessible from an internal network.

Reviewer: ```suggestion
- The resource is inaccessible. For example, if authentication is required, or if the resource is only accessible from an internal network.
```

> 3. The resource is blocked by the browser due to CORS (typically, web-fonts). 

Reviewer: ```suggestion
- The resource is blocked by the browser due to CORS (typically, web fonts). 
```

> The replay being rendered on the `app.datadoghq.com` domain, the assets requests are subject to cross origin security checks by your browser. If the given asset is not authorised for the domain, your browser blocks the request.

Reviewer: ```suggestion
    - If the replay is being rendered on the `app.datadoghq.com` domain, the assets requests are subject to cross-origin security checks by your browser. If the given asset is not authorized for the domain, your browser blocks the request.
```

> The fix is thus to allow `app.datadoghq.com through` the `[Access-Control-Allow-Origin][5]` header for any font or image assets your website depends upon, ensuring these resources are accessible for the replay.

Reviewer: ```suggestion
    - The fix is to allow `app.datadoghq.com` through the `[Access-Control-Allow-Origin][5]` header for any font or image assets your website depends on, ensuring these resources are accessible for the replay.
```

> To learn more about Cross Origin Resource Sharing, see the [MDN Web Docs article][6].

Reviewer: ```suggestion
To learn more about Cross-Origin Resource Sharing, see the [MDN Web Docs article][6].
```

### CSS rules not properly applied / mouse hover not replayed 

> Contrary to fonts & images, the recorder will try to bundle the various CSS rules applied as part of the recording data, leveraging the [CSSStyleSheet][7] interface. If this is not possible, it will fallback to recording the links to the CSS files.

Reviewer: ```suggestion
Unlike fonts and images, the recorder tries to bundle the various CSS rules applied as part of the recording data, leveraging the [CSSStyleSheet][7] interface. If this is not possible, it falls back to recording the links to the CSS files.
```

Note that for proper mouse hover support, the CSS rules must be accessible via the CSSStyleSheet interface.

> If the stylesheets are hosted on a different domain than the web page, access to the CSS rules is subject to cross origin security checks by the browser and the browser must be instructed to load the stylesheet leveraging CORS via the [crossorigin][8] attribute.

Reviewer: ```suggestion
If the stylesheets are hosted on a different domain than the web page, access to the CSS rules is subject to cross-origin security checks by the browser, and the browser must be instructed to load the stylesheet leveraging CORS using the [crossorigin][8] attribute.
```

For example, if your application is on the `example.com` domain and depends on a CSS file on `assets.example.com` via a link element, the `crossoriring` attribute should be set to `anonymous`, unless credentials are required:

``` html
<link rel="stylesheet" crossorigin="anonymous"
      href="https://assets.example.com/style.css”>

> ```

Reviewer: there is something broken with this code example that is affecting the rest of the page. I can look once I finish the general review

Author: what do you mean?

> Additionally, the `example.com` domain must be authorised by `assets.example.com` to load it’s resource by properly setting the `[Access-Control-Allow-Origin][5]` header.

Reviewer: ```suggestion
Additionally, authorize the `example.com` domain in the `assets.example.com`. This allows the assets file to load the resource by setting the `[Access-Control-Allow-Origin][5]` header.
```

> ## Frequently asked questions

Reviewer: We don't have FAQs in docs pages because we find that it has poor SEO, instead we should add this content where it makes the most sense in the rest of the document. The "How it works" section should be moved up under the introduction "What is Session Replay?" section. The performance impact could also be added into the introduction, and then the availability section can be added to the configuration section.

Author: This is not a page that will be indexed, we'll integrate Session Replay docs into the full docs once it's GA
